### PR TITLE
JNG-4741 Incremental code generation

### PR DESCRIPTION
**Pull request by robertcsakany**
<table>
<td>
  <a href="https://blackbelt.atlassian.net/browse/JNG-4741" title="JNG-4741" target="_blank">StoryJNG-4741</a>  Incremental template generation
  </td></table>
  <br />
 

JNG-4741 Incremental code generation

### CODE REVIEW

**Review by noherczeg — APPROVED** [review]

File: src/main/java/hu/blackbelt/judo/generator/commons/ChecksumUtil.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import java.security.NoSuchAlgorithmException;
+
+public class ChecksumUtil {
+
+    public static String getMD5(byte[] input) {
+        try {
+            MessageDigest md = MessageDigest.getInstance("MD5");
+            return format(md.digest(input));
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("Could notnget Md5 sum", e);
```
Comment (bencegelei) [inline src/main/java/hu/blackbelt/judo/generator/commons/ChecksumUtil.java:39]: > notnget

not get

File: src/main/java/hu/blackbelt/judo/generator/commons/ChecksumUtil.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            MessageDigest md = MessageDigest.getInstance("MD5");
+            try (BufferedInputStream in = new BufferedInputStream((new FileInputStream(path.toFile())));
+                 DigestOutputStream out = new DigestOutputStream(OutputStream.nullOutputStream(), md)) {
+                in.transferTo(out);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            return format(md.digest());
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("Could not nget Md5 sum", e);
```
Comment (bencegelei) [inline src/main/java/hu/blackbelt/judo/generator/commons/ChecksumUtil.java]: > nget

get

File: src/main/java/hu/blackbelt/judo/generator/commons/ChecksumUtil.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        }
+    }
+
+    private static String format(byte[] md) {
+        BigInteger number = new BigInteger(1, md);
+        String hashtext = number.toString(16);
+        // Now we need to zero pad it if you actually want the full 32 chars.
+        while (hashtext.length() < 32) {
+            hashtext = "0" + hashtext;
+        }
```
Comment (bencegelei) [inline src/main/java/hu/blackbelt/judo/generator/commons/ChecksumUtil.java:64]: >        while (hashtext.length() < 32) {
>            hashtext = "0" + hashtext;
>        }

This looks funny:D

File: src/main/java/hu/blackbelt/judo/generator/commons/GeneratorFileEntry.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o instanceof GeneratorFileEntry) {
+            GeneratorFileEntry fileEntry = (GeneratorFileEntry) o;
+            return this.getPath().equals(fileEntry.getPath()) && this.getChecksum().equals(fileEntry.getChecksum());
+        }
+        return false;
+    }
```
Comment (bencegelei) [inline src/main/java/hu/blackbelt/judo/generator/commons/GeneratorFileEntry.java:76]: > getPath().hashCode(); // hashcode

> this.getPath().equals(fileEntry.getPath()) && this.getChecksum().equals(fileEntry.getChecksum()); // equals logic

Couldn't this cause a hash collision?

File: src/main/java/hu/blackbelt/judo/generator/commons/GeneratorIgnore.java [hunk 5]
```diff
@@ -20,43 +20,82 @@
...
+
+    CacheLoader<Path, List<String>> loader = new CacheLoader<Path, List<String>>() {
+        @Override
+        public List<String> load(Path key) {
+            if (key.resolve(GENERATOR_IGNORE_FILE).toFile().exists()) {
+                try {
+                    List<String> globs = Files.readAllLines(key.resolve(GeneratorIgnore.GENERATOR_IGNORE_FILE), StandardCharsets.UTF_8);
+                    return globs;
+                } catch (IOException e) {
+                    throw new RuntimeException("Could not read file: " + key.resolve(GENERATOR_IGNORE_FILE).toFile().getAbsolutePath());
```
Comment (bencegelei) [inline src/main/java/hu/blackbelt/judo/generator/commons/GeneratorIgnore.java]: > throw new RuntimeException("Could not read file: " + key.resolve(GENERATOR_IGNORE_FILE).toFile().getAbsolutePath());

Original exception could be passed to runtime exc.

File: src/main/java/hu/blackbelt/judo/generator/commons/GeneratorIgnore.java [hunk 6]
```diff
@@ -20,43 +20,82 @@
...
+        }
+        return match;
+    }
+
+
+    private boolean isChildPath(Path possibleParent, Path maybeChild) {
+        File fileToTest = maybeChild.toFile();
+        while (fileToTest != null) {
+            if (fileToTest.equals(possibleParent.toFile()))
+                return true;
```
Comment (bencegelei) [inline src/main/java/hu/blackbelt/judo/generator/commons/GeneratorIgnore.java]:
```
if (fileToTest.equals(possibleParent.toFile()))
    true;
```
Please add curly brackets

File: src/main/java/hu/blackbelt/judo/generator/commons/TemplateEvaulator.java [hunk 7]
```diff
@@ -21,12 +21,15 @@
  */
 
 import com.github.jknack.handlebars.Template;
+import com.google.common.collect.Lists;
```
Comment (bencegelei) [inline src/main/java/hu/blackbelt/judo/generator/commons/TemplateEvaulator.java]: What's this import for?

**Review by bencegelei — CHANGES_REQUESTED** [review]

File: src/main/java/hu/blackbelt/judo/generator/commons/ChecksumUtil.java (hunk 3 above)
Comment (bencegelei) [reply src/main/java/hu/blackbelt/judo/generator/commons/ChecksumUtil.java:64]: (not a change request)

Comment (robertcsakany) [reply src/main/java/hu/blackbelt/judo/generator/commons/ChecksumUtil.java:64]: Yeah its funny. Do not wanna use apache commons for lpad. Its the fastest way :)

File: src/main/java/hu/blackbelt/judo/generator/commons/GeneratorFileEntry.java (hunk 4 above)
Comment (robertcsakany) [reply src/main/java/hu/blackbelt/judo/generator/commons/GeneratorFileEntry.java:76]: Yes, thats the reason. When path matching means same entity

File: src/main/java/hu/blackbelt/judo/generator/commons/GeneratorIgnore.java (hunk 6 above)
Comment (robertcsakany) [reply src/main/java/hu/blackbelt/judo/generator/commons/GeneratorIgnore.java]: Ah, copy paste from stack overflow 🗡️

**Pull request merged**